server/src/main/java/org/opensearch/index/compositeindex/datacube/startree/builder/BaseStarTreeBuilder.java

```diff
@@ -470,20 +481,40 @@ public List<SequentialDocValuesIterator> getMetricReaders(SegmentWriteState stat
         List<SequentialDocValuesIterator> metricReaders = new ArrayList<>();
         for (Metric metric : this.starTreeField.getMetrics()) {
             for (MetricStat metricStat : metric.getMetrics()) {
+                SequentialDocValuesIterator metricReader = null;
                 FieldInfo metricFieldInfo = state.fieldInfos.fieldInfo(metric.getField());
                 if (metricFieldInfo == null) {
                     metricFieldInfo = getFieldInfo(metric.getField());
                 }
-
-                SequentialDocValuesIterator metricReader = new SequentialDocValuesIterator(
-                    fieldProducerMap.get(metricFieldInfo.name).getSortedNumeric(metricFieldInfo)
-                );
+                if (metricStat.equals(MetricStat.DOC_COUNT)) {
+                    metricReader = getDocCountMetricReader(fieldProducerMap, metricFieldInfo);
+                } else {
+                    metricReader = new SequentialDocValuesIterator(
+                        fieldProducerMap.get(metricFieldInfo.name).getSortedNumeric(metricFieldInfo)
+                    );
+                }
                 metricReaders.add(metricReader);
             }
         }
         return metricReaders;
     }
 
+    private static SequentialDocValuesIterator getDocCountMetricReader(
+        Map<String, DocValuesProducer> fieldProducerMap,
+        FieldInfo metricFieldInfo
+    ) throws IOException {
+        SequentialDocValuesIterator metricReader;
+        // _doc_count is numeric field , so we need to get sortedNumericDocValues
+        if (fieldProducerMap.containsKey(metricFieldInfo.name)) {
+            metricReader = new SequentialDocValuesIterator(
+                DocValues.singleton(fieldProducerMap.get(metricFieldInfo.name).getNumeric(metricFieldInfo))
+            );
+        } else {
```

Comment:
I think you did this if else for flush and merge distinction. Are we sure doc count cannot be null?

Reply:
This flow is triggered as part of flush. 
Since during flush we always create `_doc_count` metric field, merge will always have the field.